sourcerer-eventstore/src/main/java/org/elder/sourcerer/eventstore/EventStoreEventRepository.java

```diff
@@ -101,10 +101,56 @@ public EventStoreEventRepository(
         return readInternal(toEsStreamId(streamId), version, maxEvents, false);
     }
 
+    @Override
+    public EventRecord<T> readFirst(final String streamId) {
+        return readSingleInternal(
+                toEsStreamId(streamId),
+                new EventNumber.Exact(0),
+                false);
+    }
+
+    @Override
+    public EventRecord<T> readLast(final String streamId) {
+        return readSingleInternal(
+                toEsStreamId(streamId),
+                new EventNumber.Last$(),
+                false);
+    }
+
     private String getCategoryStreamName() {
         return "$ce-" + streamPrefix;
     }
 
+    private EventRecord<T> readSingleInternal(
+            final String internalStreamId,
+            final EventNumber eventNumber,
+            final boolean resolveLinksTo) {
+        try {
+            logger.debug(
+                    "Reading latest event from {} (in {})",
+                    internalStreamId,
+                    streamPrefix);
+
+            ReadStreamEventsCompleted res = completeReadFuture(connection.readStreamEventsBackward(
+                    internalStreamId,
+                    eventNumber,
+                    1,
+                    resolveLinksTo,
+                    null));
+
+            logger.debug("Read event from {}", internalStreamId);
+            List<Event> events = res.eventsJava();
+            if (events.isEmpty()) {
+                logger.debug("Event from {} not found", internalStreamId);
+                return null;
+            }
+
+            return fromEsEvent(events.get(0));
+        } catch (eventstore.StreamNotFoundException ex) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.elder.org/coding_rules#rule_key=squid%3AS1166)